docs/USAGE-GUIDE.md

```diff
@@ -1585,6 +1585,102 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *n
+
+Benefits of using Pre-Shared Keys include: 
+
+- Using pre-shared keys can avoid the need for public key operations. This is useful if TLS is used in  performance-constrained environments with limited CPU power.
+- Pre-shared keys may be more convenient from a key management point of view. For instance, in closed environments where the connections are  mostly configured manually in advance, it is easier to configure a PSK  than to use certificates. Another case is when the parties already have a mechanism for setting up a shared secret key, and that mechanism could be used to "bootstrap" a key for authenticating a TLS connection.
+- Pre-shared keys are more appropriate for use cases with long-lived credentials stored in hardware secure elements than PKI-based solutions which enforce expiration dates and require error-prone rotation schemes.
+
+S2N-TLS provides the following pre-shared key related calls to configure and use pre-shared keys in the connection.
+```c
+struct s2n_psk* s2n_external_psk_new();
+int s2n_psk_free(struct s2n_psk **psk);
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size);
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
+int s2n_connection_append_psk(struct s2n_connection *conn, struct s2n_psk *psk);
+int s2n_config_set_psk_mode(struct s2n_config *config, s2n_psk_mode mode);
+int s2n_connection_set_psk_mode(struct s2n_connection *conn, s2n_psk_mode mode);
+int s2n_connection_get_negotiated_psk_identity_length(struct s2n_connection *conn, uint16_t *identity_length);
+int s2n_connection_get_negotiated_psk_identity(struct s2n_connection *conn, uint8_t *identity, uint16_t max_identity_length);
+struct s2n_offered_psk* s2n_offered_psk_new();
+int s2n_offered_psk_free(struct s2n_offered_psk **psk);
+int s2n_offered_psk_get_identity(struct s2n_offered_psk *psk, uint8_t** identity, uint16_t *size);
+bool s2n_offered_psk_list_has_next(struct s2n_offered_psk_list *psk_list);
+int s2n_offered_psk_list_next(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+int s2n_offered_psk_list_reread(struct s2n_offered_psk_list *psk_list);
+int s2n_offered_psk_list_choose_psk(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+typedef int (*s2n_psk_selection_callback)(struct s2n_connection *conn, void *context,
+                                          struct s2n_offered_psk_list *psk_list);
+int s2n_config_set_psk_selection_callback(struct s2n_config *config, s2n_psk_selection_callback cb, void *context);
+```
+
+### Configuring External Pre-Shared Keys
+
+Use the following APIs to configure and use external pre-shared keys in the > =TLS1.3 connections. An external PSK is an out-of-band provisioned pre-shared key (PSK) established externally using a secure mutually agreed upon mechanism.
+
```

Comment:
We don't have a security considerations document, but you should still mention best practices here, even if you don't go too in depth.

Reply:
What specifically do you want to call out in Best Practices here? I have tried to call out the necessary ones along with the description. For ex: using non-secret value for PSK Identity, PSK_HMAC that match cipher suite. Anything specific you are looking for me to add?

Reply:
Minimally the "one secret per client/server pair", and maybe also mention secure key generation.